oximeter/db/src/oxql/ast/table_ops/filter.rs

```diff
@@ -30,33 +30,212 @@ use std::fmt;
+    /// i.e., only has terms like `(a && b) || (c && d) || ...`.
+    ///
+    /// This method exists for the purposes of creating _independent_ pieces of
+    /// a filtering expression, each of which can be used to generate a new SQL
+    /// query run against ClickHouse. This is critical to support complicated
+    /// OxQL queries. Consider:
+    ///
+    /// ```ignore
+    /// get some_timeseries
+    ///     | filter (foo == "bar") || (timestamp > @now() - 1m && foo == "baz")
+    /// ```
+    ///
+    /// This requires fetching part of one timeseries, and all of another. One
+    /// cannot run this as a conjunction on the fields and then a query on the
+    /// measurements. It must be run in such a way to get the sets of keys
+    /// consistent with each term in the disjunction _independently_, so that
+    /// one can apply the timestamp filter to only the correct one.
+    ///
+    /// We use this method to generate the DNF, a form with only disjunctions.
+    /// Each of those is then a separate query against the fields table, where
+    /// we keep track of the keys in each. Each set of predicates and consistent
+    /// keys is then used later to fetch the measurements.
+    ///
+    /// # Notes
+    ///
+    /// There is a huge academic literature on this topic, part of the study of
+    /// formal languages and other areas theoretical computer science. These
+    /// references are mostly pretty dense and formal, though a few are really
+    /// useful. This [paper](https://www.researchgate.net/publication/220154187_A_Survey_of_Strategies_in_Program_Transformation_Systems)
+    /// is a good and accessible survey to the idea of translation systems --
+    /// it's mostly focused on programming languages and compilers, but Figures
+    /// 7-9 in particular are about DNF.
+    ///
+    /// As usual, the Wikipedia page is a reasonable overview as well,
+    /// [here](https://en.wikipedia.org/wiki/Disjunctive_normal_form). We're
+    /// using the "syntactic" DNF conversion algorithm, essentially. This
+    /// involves a recursive application of
+    /// [de Morgan's rules](https://en.wikipedia.org/wiki/De_Morgan%27s_laws),
+    /// [involution / double-negation](https://en.wikipedia.org/wiki/Involution_(mathematics)),
+    /// distributivity of [Boolean operators](https://en.wikipedia.org/wiki/Boolean_algebra#Monotone_laws),
+    /// etc.
+    pub fn simplify_to_dnf(&self) -> Result<Self, Error> {
+        let mut out = self.simplify_to_dnf_inner()?;
+        if &out == self {
+            return Ok(out);
+        }
+        // Continually apply simplifications as long as able.
+        //
+        // This makes me really nervous, so I'm adding an escape hatch that we
+        // only allow a few iterations. If we've not simplified within that,
+        // we'll just declare the expression too complicated to handle.
+        const EXPR_COMPLEXITY_LIMIT: usize = 6;
```

Comment:
So, I've been thinking about this. I don't actually think this is the right way to impose a limit, because the function is recursive! We need to pass in the current level, and just refuse to recurse more than some number of levels deep.

_This_ value prevents handling expressions like this: `a && (b || c || d .. )`, where there are more than 6 items in the right sub-expression. Each one requires a separate call to `simplify_to_dnf()` to resolve, but at the same layer of the recursive call stack. _That_ I agree can be much larger, though I'm not sure by how much. The _recursive_ limit should be a separate thing, governing how deep the parse tree get when trying to simplify it. I don't have a good intuition for a limit there either. But both being O(100) seems like way more than enough, and still reasonably safe?

Reply:
Yeah, O(100) seems good.

Reply:
Welp, I've found a new way you can DOS the query parser! Just make an expression that's a few right-nested ANDs, such as `a == 0 && (a == 0 && (a == 0 && ...)))` and BOOM, exponentially long parse times. Fun.

I think that means I need to add a stricter check in the top-level `Query::new()` method, which does the parsing internally. That could do something very dumb, like check the number of `&&` or parentheses, and bail because it looks suspicious.

Reply:
😬

Reply:
I'd check what's going on with the PEG/packrat parser there. I would not expect that kind of behavior for this grammar.

Reply:
I'm also surprised. I've asked it to print debug information about the rules as it matches them. It doesn't seem obviously wrong, but it's also possible I've written them very poorly in a way that requires much more work than I expected. I'll keep digging.

Reply:
Ok, I've a solution, and a small amount more insight. The Rust crate we're using here, `peg`, is not a caching / packrat parser by default. But you can make it that way by decorating rules you expect caching to help. I don't think I quite appreciated the degree to which this helps us. When you nest 4 right-deep AND expressions, this goes from **8.16s** down to **0.1s** when we cache only one rule, the `factor()` rule that matches pretty high-level things like `!(foo == "bar")`. So I'll definitely be adding _that_!

What I don't quite get is why the rule is being called so many times at the exact same locations. Again in that 4-level nesting case, it's being called **139 times** for the most deeply-nested expression. I mean, it sort of makes sense that this would be called frequently: `factor()` itself as a rule shows up in several places, and is nested right inside itself. But this seems way beyond excessive. With caching, it's down to 7 times at most.

Another mark for writing our own parser. I'd like to defer that, since this PR is already way too big. But I'm on board.

Reply:
Added caching and new tests for complexity limits in f27906b5a